store-redshift/src/test/scala/com/socrata/util/ResultSet.scala

```
@@ -0,0 +1,15 @@
+package com.socrata.util
+
+import java.sql.ResultSet
+
+object ResultSet {
+
+  def extract[T](res: ResultSet)(f: ResultSet => T): BufferedIterator[T] = {
```

This actually won't work. You need to accept a fn from BufferedIterator.
The resultset will be closed outside of the using block, and so the bufferediterator will be empty.

---

Perhaps? In our case it's always used from a managed resource context, and we immediately collect via a headOption. 
Seems better to leave this open and let the caller decide how they want to process, I don't see a need to make this implicitly more eager.

---

then why not only expose `headOption`?

---

```
2 files changed, 5 insertions(+), 2 deletions(-)
store-redshift/src/test/scala/com/socrata/service/QueryService.scala | 2 +-
store-redshift/src/test/scala/com/socrata/util/ResultSet.scala       | 5 ++++-

modified   store-redshift/src/test/scala/com/socrata/service/QueryService.scala
@@ -19,7 +19,7 @@ class QueryService
     Using.resource(dataSource.getConnection) { conn =>
       Using.resource(conn.prepareStatement(s"""select count(*) from "$tableName";""")) { stmt =>
         Using.resource(stmt.executeQuery()) { resultSet =>
-          val rowcount: Option[Long] = extract(resultSet)(rs => rs.getLong(1)).headOption;
+          val rowcount: Option[Long] = extractHeadOption(resultSet)(rs => rs.getLong(1));
           rowcount match {
             case Some(count) => Log.info(s"Table $tableName has $count records")
             case _ => Log.info(s"Unable to get record count for table $tableName")
modified   store-redshift/src/test/scala/com/socrata/util/ResultSet.scala
@@ -4,7 +4,10 @@ import java.sql.ResultSet
 
 object ResultSet {
 
-  def extract[T](res: ResultSet)(f: ResultSet => T): BufferedIterator[T] = {
+  def extractHeadOption[T](res: ResultSet)(f: ResultSet => T): Option[T] = {
+    extract(res)(f).headOption
+  }
+  private def extract[T](res: ResultSet)(f: ResultSet => T): BufferedIterator[T] = {
     new Iterator[T] {
       def hasNext = res.next()
 
```

---

something like this.

---

otherwise we're exposing a dangerous primitive that'll fail silently when (innocuously) used in the wrong context

---

I mean yea sure more specific extract methods can't hurt. I'm fine with adding more. I don't see or have any problem using iterators and leaving the safety of a resultset to the caller.

---

Returning a BufferedIterator is dangerous since returning the value will produce a silent bug.
I have to just know that this bufferediterator is not valid past the using block.